db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/TriggerExecutionContext.java

```diff
@@ -788,13 +788,14 @@ public boolean hasGeneratedColumn() {
         return hasGeneratedColumn;
     }
 
-    public boolean needsTemporaryConglomerate() {
-        // TODO:  Can we detect when a DML statement and associated
-        //        triggers needs no temporary conglomerate?
-        //        The persisted Dataset may not always be accessible,
-        //        for example in a nested loop join iterator,
-        //        so currently we still need it.
-        return true;
-        //return hasGeneratedColumn || !fromSparkExecution;
+    public boolean needsTemporaryConglomerate(boolean nestedTrigger) {
```

Comment:
This comment is not related to CR, but just to get better understanding of triggers logic.

Could you please explain why do we need temporary conglomerate for triggers, and what is the life-cycle of the temporary conglomerate? Are there other cases besides triggers where temp. conglomerates are required?

Reply:
When running on control, the temporary trigger rows can simply be cached in memory (see rowArray in TriggerRowHolderImpl) because the entire execution runs in a single process on the same node (even if using multiple threads).  But on Spark, the triggering statement (insert, update, delete) runs in different spark executors on different nodes.  There really is no good way to collect all of these rows together in a cache managed by Splice (maybe you could do it via Zookeeper, but perhaps this would not perform well).  Using a spark Dataset makes sense, but requires that the trigger actions be invoked from the OlapServer.  For trigger actions marked as OLTP, they may cause trigger statements to be invoked from the executors, which do not have access to the Dataset collected from the OlapServer.  So, the original implementation creates a temporary conglomerate to store the rows.  Currently only spark uses the temporary conglomerate, though code-wise OLTP execution could also use it, though for performance reasons and to avoid potential bugs due to extra complexity, we only uses the in-memory cache on OLTP.  If we go above the OLTP execution threshold (1 million rows, bumped up to 10 million rows by this PR), the trigger execution is rolled back by throwing a ResubmitDistributedException and we re-run it in spark.  The temporary trigger conglomerate is always removed via TriggerRowHolderImpl.close -> dropConglomerate.  I don't know if any other features use a temporary conglomerate.

Reply:
Would this be a good comment in the code?

Reply:
@carolp-503 I've added this comment, slightly modified, to the top of TriggerRowHolderImpl.

Reply:
Thanks Mark for taking the time explaining this.